### PR TITLE
Added option to specify x and y components of camera focal length.

### PR DESCRIPTION
**Pull request by carlosmccosta**

### CODE REVIEW

File: gazebo_plugins/src/gazebo_ros_camera_utils.cpp [hunk 1]
```diff
@@ -175,13 +175,34 @@ void GazeboRosCameraUtils::Load(sensors::SensorPtr _parent,
...
-  if (!this->sdf->HasElement("focalLength"))
+  if (this->sdf->HasElement("focalLength"))
   {
-    ROS_DEBUG("Camera plugin missing <focalLength>, defaults to 0");
-    this->focal_length_= 0;
+    this->focal_length_x_ = this->sdf->Get<double>("focalLength");
+    this->focal_length_y_ = this->focal_length_x_;
+  }
+
+  if (!this->sdf->HasElement("focalLength_x"))
```
Comment (ffurrer) [inline gazebo_plugins/src/gazebo_ros_camera_utils.cpp]: be consistent with the naming `focalLength_x` vs `focalLengthX` and make sure you only output debug messages if none of the parameters are set.

File: gazebo_plugins/src/gazebo_ros_camera_utils.cpp [hunk 2]
```diff
@@ -472,14 +493,15 @@ void GazeboRosCameraUtils::Init()
...
   {
-    this->focal_length_ = computed_focal_length;
+    this->focal_length_x_ = computed_focal_length;
+    this->focal_length_y_ = computed_focal_length;
   }
   else
   {
     // check against float precision
-    if (!gazebo::math::equal(this->focal_length_, computed_focal_length))
+    if (!gazebo::math::equal(this->focal_length_x_, computed_focal_length))
```
Comment (ffurrer) [inline gazebo_plugins/src/gazebo_ros_camera_utils.cpp:510]: you probably want to this for the y focal_length as well

Comment (ffurrer) [inline gazebo_plugins/src/gazebo_ros_camera_utils.cpp:510]: ok, nevermind

File: gazebo_plugins/src/gazebo_ros_camera_utils.cpp (hunk 1 above)
Comment (carlosmccosta) [inline gazebo_plugins/src/gazebo_ros_camera_utils.cpp]: Corrected sdf parameters names.
In relation to the warnings, it only outputs msgs if the focalLength and focalLengthX / focalLengthY aren't specified.

Thanks for the feedback :)

**Comment by hsu** [comment]
@carlosmccosta  Thank you, PR looks good. Do you have an example or test scenario for this PR?

**Comment by carlosmccosta** [comment]
Projection mapping example:
https://github.com/inesc-tec-robotics/gazebo_projection_mapping

Gazebo world:
https://github.com/inesc-tec-robotics/gazebo_projection_mapping/blob/hydro-devel/worlds/welding.world#L443

**Comment by carlosmccosta** [comment]
I added a testing world with a chess board in:
https://github.com/inesc-tec-robotics/gazebo_projection_mapping/blob/hydro-devel/worlds/camera_intrinsics.world

Some images showing the affect of increasing and decreasing each camera intrinsic parameters are available at:
https://github.com/inesc-tec-robotics/gazebo_projection_mapping/tree/hydro-devel/docs/camera_intrinsics

If you need further improvements let me known :)
Have a nice day :)

**Comment by j-rivero** [comment]
@osrf-jenkins run test please

**Comment by carlosmccosta** [comment]
Good afternoon,

The error reported is due to this latter commit:
https://github.com/ros-simulation/gazebo_ros_pkgs/commit/dacd87275dda7bec350cb4992dfa868b141c1b52#diff-54a33e23f1a882810606eba53a485312R497

That overwrote this change:
https://github.com/ros-simulation/gazebo_ros_pkgs/pull/374/commits/35d585fc52089c1dd948eb186b8558e2e3964f1b#diff-54a33e23f1a882810606eba53a485312R512

And reverted the update of the field name from this->focal_length_x_ to this->focal_length_

To fix this, either
- The pull request is merged and a new commit is made to remove the changes that reverted the this->focal_length_x_ member name
- I apply the changes of this pull request into the current head of the jade-devel branch and create another pull request or overwrite the history of this pull request

How do you want to proceed?

Have a nice day :)

**Comment by j-rivero** [comment]
> How do you want to proceed?

Thanks for the quick answer Carlos. I would prefer not to break the gazebo7 build in any moment in the current -devel branches, so I think that the best option is the second one that you proposed: update this PR with latest jade-devel changes. I really don't mind about creating a new pull request or overwrite this one, the one that works better for you.

Hace a good day you too :)

**Comment by carlosmccosta** [comment]
Good morning,

I rebased this PR commit on top of the current jade-devel branch and rewrote the carlosmccosta:jade-devel git history.
This way there is a clean merge of a single commit on top of ros-simulation:jade-devel.
All CI tests passed. If you need further corrections / improvements let me know :)

Have a nice day :)

**Comment by carlosmccosta** [comment]
Rebase of this PR for melodic-devel in #767

**Comment by ahcorde** [comment]
No activity since 2018 and targeting and unsupported ROS version

**Pull request closed without merging**